### DISCUSSION TITLE
Horrible TypeScript experience

### DISCUSSION THREAD

**Discussion by czebe** [statelyai/xstate#2163] [Q&A]
There's not much documentation around TypeScript and even less examples.

Ran into some issues with types:
- states, actions, guards, services split into separate files -> it's super hard to figure out and supply all necessary types
- guard `meta` properties are conflicting
- events custom data are not recognized from the Events type
- `assign` type issue (this was discussed before)
- state schema (as shown in the docs) fails `eslint` with default `@typescript-eslint config`

Ended up combining everything into a single file. It's somewhat better, but now it's a super huge file, very hard to navigate.
(guard meta is still causing `ts-jest` to fail)

Perhaps it's natural for someone to provide a complex example that passes linting, tsc and ts-jest.

**Comment by rjdestigter** [comment]
I highly recommend using https://github.com/mattpocock/xstate-codegen 

Combined with that I write my state machine definitions completely JSON serializable _(first argument passed to `createMachine`)_ The implementation of guards, actions, and services passed as a second argument is primarily imported. This does mean that when defining action using `assign_ for example you have to provide it with the type information. E.g.:

```
assign<MyMachineContext, MyMachineEvents>( .. )
```

If you choose not to import them but define them on the spot then type inference will work though.

**Reply by czebe** [reply]
Thanks!! I tried that tool, but it failed to produce anything for me. Probably errored out. I’ll give it another try

**Reply by ChrisShank** [reply]
`createModel` has also been a gamechanger for me with typings in xstate!

https://github.com/davidkpiano/xstate/blob/master/packages/core/CHANGELOG.md#4170

**Comment by threehams** [comment]
Is there a real-world example out there of an "ideal" setup and experience with TypeScript?

After a day of messing around with xstate, typestates, createModel, createSchema, and xstate-codegen on a very simple state machine, I can't find a combination that does all of the following:
- catches obvious mistakes in the machine definition itself, like invalid states or targets
- provides correct types for both `value` and `matches`
- only accepts valid events for `send`

xstate itself can deal with value/matches, but nearly everything in the definition itself is `any` or `string`.
xstate-codegen validates the machine config and `send`, but doesn't type either `matches` or `value` because it doesn't support typestates.

**Reply by siman** [reply]
+1 for a real-world example in TypeScript

**Reply by tomByrer** [reply]
I suspect the next major version will be more TypeScripty.
